www/_blog/2022-06-30-error-handling.mdx

```diff
@@ -0,0 +1,407 @@
+
+In the standard library [VarError](https://doc.rust-lang.org/std/env/enum.VarError.html) (for reading environment variables) and [ParseIntError](https://doc.rust-lang.org/std/num/struct.ParseIntError.html) (for parsing a string slice as a integer) are different errors. When we interact them we need to differentiate between the types because they have different properties and different stack sizes. To build a combination of them we could build a sum type using a enum. Alternatively we can use dynamically dispatched traits which handled varying stack sized items and other type information.
+
+Using the above mentioned try syntax (`?`) we can convert the the above errors to be dynamically dispatched. This makes it easy to handle different errors without building enums to combine errors.
+
+```rust
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let key = std::env::var("NUMBER_IN_ENV")?;
+    let number = key.parse::<i32>()?;
+    println!("\"NUMBER_IN_ENV\" is {}", number);
+    Ok(())
+}
+```
+
+[(main function can return Result)](https://github.com/rust-lang/rust/pull/49162)
+
+While this is a easy way to handle errors, it isn't easy to differentiate between the types and can make handling errors in libraries hard. More information on this later.
+
+### Methods on Result
+
+Result and Option contains many useful functions. Here are some of my commonly used:
+
+##### [Result::map](https://doc.rust-lang.org/std/result/enum.Result.html#method.map)
+
+This maps or converts the Ok value if it exists. This can be more concise than using the `?` operator.
+
+```rust
+fn string_to_plus_one(s: &str) -> Result<i32, std::num::ParseIntError> {
+    s.parse::<i32>().map(|num| num + 1)
+}
+```
+
+##### [Result::ok](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok)
+
+Useful for converting Results to Options
+
+```rust
+assert_eq!(Ok(2).ok(), Some(2));
+assert_eq!(Err("err!").ok(), None);
+```
+
+### [Option::ok_or_else](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else)
+
+Useful for going the other way in converting from Options to Results
+
+```rust
+fn get_first(vec: &Vec<i32>) -> Result<&i32, NotInVec> {
+    vec.first().ok_or_else(|| NotInVec)
+}
+```
+
+### Error handling for iteration 
+
+Using results in iterator chains can be a little confusing. Luckily Result implements collect. We can use this to short circuit a iterator if a error occurs. In the following, if all the `parse`s succeed then we get collected vec of numbers result. If one fails then it instead returns a Result with the failing Err.

```

Comment:
'short circuit a iterator if a error occurs' -> 'short circuit an iterator if an error occurs'.  I think I am starting to sound like clippy in pedantic mode :-)